### PR TITLE
Expose session naming to console users

### PR DESCRIPTION
**Pull request by wvu**
# WIP

**TODO:** Continue refactoring other options to take names instead of just IDs.

- [x] `sessions -n name -i id`
- [x] `sessions`
- [x] See name(s)
- [x] `sessions -v`
- [x] See name(s)

### CODE REVIEW

**Comment by busterb** [comment]
Nice! Oldie but a goodie.

**Comment by busterb** [comment]
Any opinion @OJ on how this might mesh with guids later? Should we assign the name based on a guid -> name lookup table rather than having it in the session object itself?

**Comment by OJ** [comment]
I think that having the option to interact based on `index` (as we do now) is great for the UI. I think adding the support for interacting based on `name` is a good move forward as that really helps target specific sessions without having to remember the index, or constantly return to the console to re-list the existing sessions.

I also think that adding support, in some way, to use `guid` would allow for more advanced stuff, such as scripting.

As far as where it lives is concerned, I'm not sure. I think if it's an attribute that in some way labels the session it should be on the session object itself. Should there be a constraint placed on the name that it has to be unique across all sessions?

**Comment by busterb** [comment]
I kind of think that names should be unique, and if you want to use the same thing across multiple sessions, that's a label (something we don't have today). Either way, I don't think that holds this PR up at all.

**Comment by OJ** [comment]
Agreed!

File: lib/msf/ui/console/command_dispatcher/core.rb [hunk 1]
```diff
@@ -1473,6 +1476,27 @@ def cmd_sessions(*args)
       print_line
       print(Serializer::ReadableText.dump_sessions(framework, :show_extended => show_extended, :verbose => verbose, :search_term => search_term))
       print_line
+    when 'name'
+      if session_name.nil? || session_name.blank?
```
Comment (pbarry-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb]: Is just .blank sufficient here?  If nothing else, just for my own edification...  :P

Comment (wvu-r7) [reply lib/msf/ui/console/command_dispatcher/core.rb]: It used to be `session_name.strip.empty?`, so I forgot to remove the `nil` case. Thanks!

Comment (wvu-r7) [reply lib/msf/ui/console/command_dispatcher/core.rb]: (It's been hard for me to start adopting the Railsy stuff. :)

**Comment by busterb** [comment]
You mean activesupport? https://github.com/rails/rails/tree/master/activesupport 
Think of it more like Boost for Ruby.

**Comment by sempervictus** [comment]
Seems to rename all sessions though... Ignores the session index passed (in my branch anyway).

**Comment by wvu-r7** [comment]
@sempervictus: Do you have the latest commit? I rebased it to remove that behavior, since I didn't like it. Specify an index with `-i`.

**Comment by sempervictus** [comment]
Ah, sorry, disregard :-).

**Comment by wvu-r7** [comment]
Yeah... I'm a sneaky rebaser. Apologies.

**Comment by sempervictus** [comment]
For some reason that made me think of Futurama's "sneaky male" concept. :-)

**Comment by bwatters-r7** [comment]
## Testing
Commands:
```session -n testsession -i 1```
```sessions -l```
Success Criteria:
```1   testsession```
## x86 Payloads:
image
image

**Comment by bwatters-r7** [comment]
Still chasing down the failures.  They have nothing to do with the PR changes, as those errors are simply sessions that did not establish.  Logs show that the SYN packet gets sent by the target, but pcaps on the msfconsole vm do not show the packet ever arrives.  I'm still troubleshooting that one.

**Comment by bwatters-r7** [comment]
## x64 Windows
image
image

The error above was the same as before

**Comment by bwatters-r7** [comment]
I noticed above I forgot to change the exe payloads to 64 bit.  Changed those and reran.  Still worked.

**Comment by wvu-r7** [comment]
:cake:

**Comment by bwatters-r7** [comment]
# Release Notes
This PR exposes the ability to name sessions.  The option existed in the code, but there was never a way for users to do it.  Now there is. 👊 Also, I'm sorry for whoever does the write-up if the Unicode breaks everything.

**Pull request merged**